include/zephyr/instrumentation/instrumentation.h

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright 2023 Linaro
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_INSTRUMENTATION_INSTRUMENTATION_H_
+#define ZEPHYR_INCLUDE_INSTRUMENTATION_INSTRUMENTATION_H_
+
+#include <zephyr/kernel.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if !defined(__no_instrumentation__)
+#error "No toolchain support for __no_instrumentation__"
+#endif
+
+/**
+ * @brief Instrumentation event record types.
+ */
+enum instr_event_types {
+	INSTR_EVENT_ENTRY = 0,	/* Callee entry event record, followed by instr_event. */
+	INSTR_EVENT_EXIT,	/* Callee exit event record, followed by instr_event. */
+	INSTR_EVENT_PROFILE,	/* Profile events */
+	INSTR_EVENT_SCHED_IN,	/* Thread switched in scheduler event */
+	INSTR_EVENT_SCHED_OUT,	/* Thread switched out scheduler event */
+	INSTR_EVENT_NUM,	/* Add more events above this one */
+	INSTR_EVENT_INVALID	/* Invalid or no event generated after promotion */
+} __packed;
+
+/**
+ * @brief Header for the event records.
+ */
+struct instr_header {
+	/* Event type */
+	enum instr_event_types type;
+} __packed;
+
+struct instr_event_context {
+	/** Arch-specific mode indicator (thread mode, interrupt mode, etc.). */
+	uint8_t mode : 3;
+	/** CPU number. */
+	uint8_t cpu : 3;
+	/** Reserved for future use. */
+	uint8_t rsvd : 2;
```

Comment:
I may have missed it in the RFC, what future plans do you foresee for the rsvd bits here?